### DISCUSSION TITLE
How can i manual release unallocated, reserved by the OS

### DISCUSSION THREAD

**Discussion by yongzhewunan** [rabbitmq/rabbitmq-server#3614] [Other] [answered]
Hi All:
    I have a distributed RabbitMQ( Server version:3.8.9) ，one day the business send too many message to queue , then i purge the messages.
now i find one node occupy large memery,  How to free,please help me,thanks!
![111](https://user-images.githubusercontent.com/16895463/138990464-7ad83c61-3e9b-477c-bccf-a4006de9aff4.png)
![222](https://user-images.githubusercontent.com/16895463/138990469-5e833133-64bc-4847-9dfb-5d9f8925fbf1.png)

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment, marked as answer]
You can't without restarting the node. You can [try different allocator settings](https://www.rabbitmq.com/runtime.html#allocators) and/or lower `max_message_size` which by default is 512 MiB IIRC.

**Reply by yongzhewunan** [reply]
Hi michaelklishin
 thinks your answer,the message queue has been cleaned, but the memory can not free, erlang vm occupy a few,I check the rabbitmq node thread heap,It's also small. so I can not understand, where to go rabbitmq applyed unused memory? direct memory? can erlang vm  reuse it ? the large unallocated reserved by the OS memory . thinks

**Comment by michaelklishin** [comment]
FWIW I remember seeing some changes considered for Erlang 25 that would allow memory allocators to release reserved and/or allocated but unused memory back to the OS. This won't ship until May 2022.

**Reply by rayate2410** [reply]
Hi.. 

NOTE: I'm aware that this is an old thread, but didn't want to create a duplicate for the exact same observation. If required can create a new thread.

I am also facing same issue with our containerized deployment of rabbitmq where we want to limit the memory of the rabbitmq pod as per our usecase. (benchmark the memory usage and predict the future requirements and accordingly set the pod memory limit)

**Some information about the setup:**
`rabbitmq: 3.11.13`
**memory related settings: (mostly default values)**
```
vm_memory_high_watermark set to 0.4 
vm_memory_calculation_strategy = rss
background_gc_enabled = true
background_gc_target_interval = 60000
```
**Workload**: 
- 1 amqp producer producing ~85MB message/minute - 3 WebSTOMP UIs listening to this data. Fanout exchange, classic queues.
- Other smaller transaction notifications at infrequent interval.

To check the memory usage, I started checking below data,
- docker stats
- memory breakdown on management UI.

Observing similar behavior where once the RSS size has grown, it doesn't shrink much - and can see large chunk of memory under "reserved unallocated". Due to this the high watermark alarm goes off, throttling the publishing sometimes.

![image](https://github.com/rabbitmq/rabbitmq-server/assets/6076575/6c249af4-d4a0-431e-aee1-adb526652704)


I have tried to use vm_memory_calculation_strategy = allocated, and understood that with this the alarm won't go off. But the actual memory consumed by rabbitmq pod still includes the "reserved unallocated".

Now, since we are using latest RMQ with Erlang 25, can you throw some light on what memory allocator settings can be done to send unused memory back to the OS. And what would be the side-effects of such configurations on performance etc.

Thank you in advance.

**Comment by johanrhodin** [comment]
In your case it looks like you should be able to switch to the `allocated`-strategy since it is only using 5.5 GB. Did you run `rabbitmqctl force_gc`? Did it clean anything up?

**Reply by yongzhewunan** [reply]
After  I executed  rabbitmqctl force_gc command，Memory is not reduced. and It look like memory leak.

**Reply by johanrhodin** [reply]
And if you change to `allocated`?
`rabbitmqctl eval 'application:set_env(rabbit, vm_memory_calculation_strategy, allocated).'`
